node/primitives/src/lib.rs

```diff
@@ -404,4 +405,37 @@ impl SignedDisputeStatement {
 	pub fn session_index(&self) -> SessionIndex {
 		self.session_index
 	}
+
+	/// Convert a [`SignedFullStatement`] to a [`SignedDisputeStatement`]
+	///
+	/// As [`SignedFullStatement`] contains only the validator index and
+	/// not the validator public key, the public key must be passed as well,
+	/// along with the signing context.
+	///
+	/// This does signature checks again with the data provided.
+	pub fn from_backing_statement(
+		backing_statement: &UncheckedSignedFullStatement,
+		signing_context: SigningContext,
+		validator_public: ValidatorId,
+	) -> Result<Self, ()> {
+		let (statement_kind, candidate_hash) = match backing_statement.unchecked_payload() {
+			Statement::Seconded(candidate) => (
+				ValidDisputeStatementKind::BackingSeconded(signing_context.parent_hash),
+				candidate.hash(),
+			),
+			Statement::Valid(candidate_hash) => (
+				ValidDisputeStatementKind::BackingValid(signing_context.parent_hash),
+				*candidate_hash,
+			),
+		};
+
+		let dispute_statement = DisputeStatement::Valid(statement_kind);
+		Self::new_checked(
+			dispute_statement,
+			candidate_hash,
+			signing_context.session_index,
+			validator_public,
+			backing_statement.unchecked_signature().clone(),
+		)
+	}
```

Comment:
Thinking again of this, I am not sure including the actual public key gains us anything here. Yes we can then verify the signature without a `SessionInfo`, but for checking that the key was indeed an actual valid validator key we still depend on `SessionInfo`. And in the end this is what matters - fine we have a valid signature, but everyone can provide a valid signature with a self chosen key. 

I would argue that, by checking against the included key, instead of a validator index, we in fact provide less safety. Because previously you had to look up that key and the only sane way to do that was to query the `SessionInfo`. Now you can check the signature without that and say "This is valid.", when in fact you don't actually know, you only know it has been signed with some key.

So yes it is cumbersome (especially in a pure and completely type checked way) to check those signatures, because they are based on session keys, but we need to deal with it. Maybe we can have some cryptographic proof, that a given key belongs to a validator in a particular session?

Reply:
But even if, that would be based on keys again and those would need to be checked again for their authenticity. So yeah, we need to query the chain, it is after all the source of truth.

Reply:
Having said that, I think we should get this PR merged ASAP to avoid further merge conflicts. My concerns don't actually relate to this PR specifically.

Reply:
Not a hlll I will die on, but 99% of the time when you're using a signed statement, you're updating some metadata which is based on an underlying `[ValidatorId]`. This necessitates validating that `statement.public() == validators[index]` and is only an obvious check when the `ValidatorIndex` _isn't_ part of `statement` already.

Our use-cases of backing, bitfields, approvals, and disputes all meet this criterion.

Reply:
That's a good point! I still like the other API more though: Ok, the smart constructor is not perfect, but it will catch almost all mistakes one can make (especially the most realistic ones, like just forgetting to check at all). The `SignedDisputeStatement` API requires us to carry around a tuple `(SignedDisputeStatement, ValidatorIndex)` which could be totally made up and is also rather inconvenient to use.

Anyhow, let's leave it as is for now. I was just wondering, if we kept a compatible implementation, whether we could skip the redundant signature checks (`from_backing_statement(statement.as_unchecked()`) calls).